service/collector/nat.go

```diff
@@ -0,0 +1,270 @@
+}
+
+type natCache struct {
+	cache *cache.StringCache
+}
+
+type natInfoResponse struct {
+	vpcs map[string]vpcInfo
+}
+
+type vpcInfo struct {
+	natGatewaysByZone map[string]float64
+}
+
+func NewNAT(config NATConfig) (*NAT, error) {
+	if config.Helper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Helper must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.InstallationName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.InstallationName must not be empty", config)
+	}
+
+	v := &NAT{
+		// AWS operator creates at this moment one NAT for each private subnet (node
+		// pool). As clusters are not created nor changed so often, and the process
+		// can take around 20 minutes, 30 minutes for the cache expiration is a
+		// reasonable value.
+		cache:  newNATCache(time.Minute * 30),
+		helper: config.Helper,
+		logger: config.Logger,
+
+		installationName: config.InstallationName,
+	}
+
+	return v, nil
+}
+
+func newNATCache(expiration time.Duration) *natCache {
+	cache := &natCache{
+		cache: cache.NewStringCache(expiration),
+	}
+
+	return cache
+}
+
+func (n *natCache) Get(accID string) (*natInfoResponse, error) {
+	var c natInfoResponse
+	raw, ok := n.cache.Get(prefixNATcacheKey + accID)
+	if ok {
+		err := hprose.Unserialize(raw, c, true)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return &c, nil
+}
+
+func (n *natCache) Set(accID string, content natInfoResponse) error {
+	contentSerialized, err := hprose.Serialize(content, true)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	n.cache.Set(prefixNATcacheKey+accID, contentSerialized)
+
+	return nil
+}
+
+func (v *NAT) Collect(ch chan<- prometheus.Metric) error {
+	reconciledClusters, err := v.helper.ListReconciledClusters()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	awsClientsList, err := v.helper.GetAWSClients(reconciledClusters)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, item := range awsClientsList {
+		awsClients := item
+
+		err := v.collectForAccount(ch, awsClients)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (v *NAT) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- natDesc
+	return nil
+}
+
+func (v *NAT) collectForAccount(ch chan<- prometheus.Metric, awsClients clientaws.Clients) error {
+	accountID, err := v.helper.AWSAccountID(awsClients)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var natInfo *natInfoResponse
+	// Check if response is cached
+	natInfo, err = v.cache.Get(accountID)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	//Cache empty, getting from API
+	if natInfo == nil || natInfo.vpcs == nil {
+		natInfo, err = getNatInfoFromAPI(accountID, awsClients)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	if natInfo != nil {
+		err = v.cache.Set(accountID, *natInfo)
```

Comment:
Doesn't this effectively prevent eviction of a cached item forever if it was found from cache in the first place? How does it get evicted / refreshed from AWS?

Reply:
Damn it you are right, I had it inside the "cache empty" conditional but I moved later trying to make it easier to read 🤦